src/entities/options/option-type.ts

```diff
@@ -0,0 +1,181 @@
+import type { Address } from 'viem';
+import {
+  bytesToBigInt,
+  encodeAbiParameters,
+  keccak256,
+  pad,
+  sliceHex,
+  toBytes,
+} from 'viem';
+import type { PublicClient } from '@wagmi/core';
+import type { SimulatedTxRequest } from '../../types';
+import type { Trader } from '../trader/base-trader';
+import { ClearinghouseContract } from '../contracts/clearinghouse';
+
+export interface OptionTypeInfo {
+  underlyingAsset: Address;
+  underlyingAmount: bigint;
+  exerciseAsset: Address;
+  exerciseAmount: bigint;
+  exerciseTimestamp: number;
+  expiryTimestamp: number;
+}
+
+type OptionTypeArgs =
+  | {
+      publicClient: PublicClient;
+      optionInfo: OptionTypeInfo;
+      tokenId?: undefined;
+    }
+  | {
+      publicClient: PublicClient;
+      tokenId: bigint;
+      optionInfo?: undefined;
+    };
+
+export class OptionType {
+  public publicClient: PublicClient;
+  public clearinghouse: ClearinghouseContract;
+
+  // Specific to OptionTypes
+  public optionInfo?: OptionTypeInfo;
+
+  // Specific to Options (Exercisable Options + Redeemable Claims)
+  private _tokenId?: bigint;
+  private _tokenType?: number;
+
+  constructor({ optionInfo, tokenId, publicClient }: OptionTypeArgs) {
+    this.publicClient = publicClient;
+    this.clearinghouse = new ClearinghouseContract({ publicClient });
+    // if passed option info directly, we can initialize immediately
+    if (optionInfo !== undefined) {
+      this.optionInfo = optionInfo;
+      void this.init();
+    } else {
+      // otherwise if the ID is the only parameter, we must derive the rest asynchronously
+      this._tokenId = tokenId;
+      void this.init(tokenId);
+    }
```

Comment:
Instead of letting the promise fly in the void, potentially changing the state of the instance, it would be better to make the `constructor` `private` – hence not callable, and create a static `from` or `initialize` method that would be asynchronous and let the consumer control the flow of async operations.

Reply:
Ah, how could I forget, I love that pattern - its much more ergonomic. Cheers